beacon_node/beacon_chain/src/state_advance_timer.rs

```diff
@@ -227,19 +230,62 @@ async fn state_advance_timer<T: BeaconChainTypes>(
 
                 // Prepare proposers so that the node can send payload attributes in the case where
                 // it decides to abandon a proposer boost re-org.
-                if let Err(e) = beacon_chain.prepare_beacon_proposer(current_slot).await {
-                    warn!(
-                        log,
-                        "Unable to prepare proposer with lookahead";
-                        "error" => ?e,
-                        "slot" => next_slot,
-                    );
-                }
+                let proposer_head = beacon_chain
+                    .prepare_beacon_proposer(current_slot)
+                    .await
+                    .unwrap_or_else(|e| {
+                        warn!(
+                            log,
+                            "Unable to prepare proposer with lookahead";
+                            "error" => ?e,
+                            "slot" => next_slot,
+                        );
+                        None
+                    });
 
                 // Use a blocking task to avoid blocking the core executor whilst waiting for locks
                 // in `ForkChoiceSignalTx`.
                 beacon_chain.task_executor.clone().spawn_blocking(
                     move || {
+                        // If we're proposing, clone the head state preemptively so that it isn't on
+                        // the hot path of proposing. We can delete this once we have tree-states.
+                        if let Some(proposer_head) = proposer_head {
+                            if let Some(proposer_state) = beacon_chain
```

Comment:
I wonder if it would be beneficial to dtop the `block_production_state` above this line. It would prevent us from holding the old state in memory whilst we load the new one.

The downside would be that there are some times where we clear the `block_production_state` when we don't need to (i.e. the state is not in the snapshot cache but would be useful for the next proposal).

Reply:
Good idea. I've added that in [6d5d558](https://github.com/sigp/lighthouse/pull/4925/commits/6d5d558a35c6863cf263f535805a1fc3d83488b2). It will still hold 2 states in memory if the beacon node has two proposals two slots in a row on the same head block, but I think this is reasonably obscure enough to be tolerated until we get tree-states. I'll try to deploy that change on Holesky in a few hours so I can confirm it didn't regress on the perf front.